### PR TITLE
add savedBattlerAttacker and stack for saved target/attacker

### PR DESCRIPTION
**Pull request by ghoulslash**
Several effects have had issues overwriting gBattlerAttacker/gBattlerTarget in the past and present. We should move away from backing up battler IDs into other id slots by using savetarget/saveattacker. So I added a stack for redundancy. An example of one possible issue would be backing up a battler id to saved battler and then having a chaining effect that also saved a different battler id to saved battler. The top-most call stack would restore the incorrect battler id.

### CODE REVIEW

File: include/battle.h [hunk 1]
```diff
@@ -698,7 +698,10 @@ struct BattleStruct
     u8 magnitudeBasePower;
     u8 presentBasePower;
     u8 roostTypes[MAX_BATTLERS_COUNT][2];
-    u8 savedBattlerTarget;
+    u8 savedBattlerTarget[10];
+    u8 savedBattlerAttacker[10];
+    u8 savedTargetCount;
```
Comment (DizzyEggg) [inline include/battle.h:703]: Just to save some space I guess savedTargetCount and savedAttackerCount could be split into one byte with 4 bits. (This is optional though)

Also 10 bytes for each saved battler seems a bit overkill, I think 5 should be fine.

File: src/battle_script_commands.c [hunk 2]
```diff
@@ -15474,6 +15465,47 @@ static void Cmd_callnative(void)
 }
 
 // Callnative Funcs
+
+static void SaveBattlerTarget(u32 battler)
+{
+    gBattleStruct->savedBattlerTarget[gBattleStruct->savedTargetCount++] = battler;
```
Comment (DizzyEggg) [inline src/battle_script_commands.c:15471]: I would add some check that savedTargetCount doesn't overflow with the array count of savedTargets. Same with attacker

File: src/battle_script_commands.c [hunk 3]
```diff
@@ -15474,6 +15465,47 @@ static void Cmd_callnative(void)
...
+{
+    NATIVE_ARGS();
+    SaveBattlerTarget(gBattlerTarget);
+    gBattlescriptCurrInstr = cmd->nextInstr;
+}
+
+void BS_RestoreTarget(void)
+{
+    NATIVE_ARGS();
+    gBattleStruct->savedTargetCount--;
```
Comment (DizzyEggg) [inline src/battle_script_commands.c:15489]: I would add a similar check here

**Review by DizzyEggg — CHANGES_REQUESTED** [review]
Good changes, should be merged soon.

**Comment by ghoulslash** [comment]
Added requisite safety checks. Also added debug warning statements as failsafe checks. Added a final safety check in MOVEEND_CLEAR_BITS to ensure all saved attackers/targets were restored within a given turn. Ready for re review

**Comment by mrgriffin** [comment]
> Also added debug warning statements as failsafe checks.

Good idea! Would it be possible to turn them into errors in tests? I think something like:
```c
#if TESTING
Test_ExitWithResult(TEST_RESULT_ERROR, "BS_RestoreTarget attempting to restore an empty target!");
#else
DebugPrintfLevel(MGBA_LOG_WARN, "BS_RestoreTarget attempting to restore an empty target!");
#endif
```
With a:
```c
#if TESTING
#include "test/test.h"
#endif
```
Ought to work. Annoyingly `Test_ExitWithResult` only exists if `TESTING` is defined, so we can't avoid the preprocessor...

Maybe in the future we should use `AGB_ASSERT` for these sorts of things, and then make that responsible for calling `Test_ExitWithResult` if `TESTING` is set (and doing the current behavior of either aborting or doing nothing if not, depending on `NDEBUG`).

**Comment by ghoulslash** [comment]
> > Also added debug warning statements as failsafe checks.
> 
> Good idea! Would it be possible to turn them into errors in tests? I think something like:
> 
> ```c
> #if TESTING
> Test_ExitWithResult(TEST_RESULT_ERROR, "BS_RestoreTarget attempting to restore an empty target!");
> #else
> DebugPrintfLevel(MGBA_LOG_WARN, "BS_RestoreTarget attempting to restore an empty target!");
> #endif
> ```
> 
> With a:
> 
> ```c
> #if TESTING
> #include "test/test.h"
> #endif
> ```
> 
> Ought to work. Annoyingly `Test_ExitWithResult` only exists if `TESTING` is defined, so we can't avoid the preprocessor...
> 
> Maybe in the future we should use `AGB_ASSERT` for these sorts of things, and then make that responsible for calling `Test_ExitWithResult` if `TESTING` is set (and doing the current behavior of either aborting or doing nothing if not, depending on `NDEBUG`).

good suggestion. Just pushed these changes. Also added to the checks in MOVEEND_CLEAR_BITS. In my project I use a lot of asserts, but warnings may be better because assertion fails just shut down mgba. But discussion on that is probably better on discord.
Anyway, ready for review again

**Comment by ghoulslash** [comment]
This keeps failing with `[Test_ExitWithResult](error: implicit declaration of function 'Test_ExitWithResult' [-Werror=implicit-function-declaration])` but builds locally

**Comment by AlexOn1ine** [comment]
> This keeps failing with `[Test_ExitWithResult](error: implicit declaration of function 'Test_ExitWithResult' [-Werror=implicit-function-declaration])` but builds locally

It doesn't build with `make -jN check`. I think you forgot to include the test runner in battle scripts commands

**Comment by ghoulslash** [comment]
> > This keeps failing with `[Test_ExitWithResult](error: implicit declaration of function 'Test_ExitWithResult' [-Werror=implicit-function-declaration])` but builds locally
> 
> It doesn't build with `make -jN check`. I think you forgot to include the test runner in battle scripts commands

Even after adding that make check fails

**Comment by AlexOn1ine** [comment]
> > > This keeps failing with `[Test_ExitWithResult](error: implicit declaration of function 'Test_ExitWithResult' [-Werror=implicit-function-declaration])` but builds locally
> > 
> > 
> > It doesn't build with `make -jN check`. I think you forgot to include the test runner in battle scripts commands
> 
> Even after adding that make check fails

My bad, sorry.

**Comment by ghoulslash** [comment]
Fixed the error by including test/test.h in battle_script_commands.c but make check them bombs with illegal op codes after a while. I think I saw an error with the sticky web/mirror armor test?

**Comment by mrgriffin** [comment]
Illegal opcodes usually means stack corruption—something overwriting an `lr` on the stack, which then causes the CPU to branch into non-code memory, (hopefully!) quickly running into bytes which don't represent a legal opcode. If you have a DACS-capable `mgba-rom-test` (which unfortunately means building it yourself or explicitly specifying a BIOS atm) then as of #3761 you should get an immediate `CRASH` instead of waiting a minute for a `TIMEOUT`.

**Comment by ghoulslash** [comment]
ready for review. I put incorporating into tests on the backburner. Will open a reminder issue upon merging

**Review by DizzyEggg — APPROVED** [review]

**Pull request merged**